// 0xsequence@2.0.10/dist/0xsequence.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var abi$1 = require('@0xsequence/abi');
var api$1 = require('@0xsequence/api');
var auth$1 = require('@0xsequence/auth');
var guard$1 = require('@0xsequence/guard');
var indexer$1 = require('@0xsequence/indexer');
var metadata$1 = require('@0xsequence/metadata');
var network$1 = require('@0xsequence/network');
var provider$1 = require('@0xsequence/provider');
var relayer$1 = require('@0xsequence/relayer');
var core$1 = require('@0xsequence/core');
var utils$1 = require('@0xsequence/utils');
var signhub$1 = require('@0xsequence/signhub');
var sessions$1 = require('@0xsequence/sessions');
var migration$1 = require('@0xsequence/migration');
var account$1 = require('@0xsequence/account');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

function _mergeNamespaces(n, m) {
  m.forEach(function (e) {
    e && typeof e !== 'string' && !Array.isArray(e) && Object.keys(e).forEach(function (k) {
      if (k !== 'default' && !(k in n)) {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  });
  return Object.freeze(n);
}

var abi__namespace = /*#__PURE__*/_interopNamespace(abi$1);
var api__namespace = /*#__PURE__*/_interopNamespace(api$1);
var auth__namespace = /*#__PURE__*/_interopNamespace(auth$1);
var guard__namespace = /*#__PURE__*/_interopNamespace(guard$1);
var indexer__namespace = /*#__PURE__*/_interopNamespace(indexer$1);
var metadata__namespace = /*#__PURE__*/_interopNamespace(metadata$1);
var network__namespace = /*#__PURE__*/_interopNamespace(network$1);
var provider__namespace = /*#__PURE__*/_interopNamespace(provider$1);
var relayer__namespace = /*#__PURE__*/_interopNamespace(relayer$1);
var core__namespace = /*#__PURE__*/_interopNamespace(core$1);
var utils__namespace = /*#__PURE__*/_interopNamespace(utils$1);
var signhub__namespace = /*#__PURE__*/_interopNamespace(signhub$1);
var sessions__namespace = /*#__PURE__*/_interopNamespace(sessions$1);
var migration__namespace = /*#__PURE__*/_interopNamespace(migration$1);
var account__namespace = /*#__PURE__*/_interopNamespace(account$1);

var abi = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [abi__namespace]);

var api = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [api__namespace]);

var auth = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [auth__namespace]);

var guard = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [guard__namespace]);

var indexer = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [indexer__namespace]);

var metadata = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [metadata__namespace]);

var network = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [network__namespace]);

var provider = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [provider__namespace]);

var relayer = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [relayer__namespace]);

const transactions = core$1.commons.transaction;

var transactions$1 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  transactions: transactions
});

var utils = /*#__PURE__*/_mergeNamespaces({
  __proto__: null,
  isValidSignature: provider$1.isValidSignature,
  isValidMessageSignature: provider$1.isValidMessageSignature,
  isValidTypedDataSignature: provider$1.isValidTypedDataSignature,
  isWalletUpToDate: provider$1.isWalletUpToDate
}, [utils__namespace]);

var core = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [core__namespace]);

var signhub = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [signhub__namespace]);

var sessions = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [sessions__namespace]);

var migration = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [migration__namespace]);

var account = /*#__PURE__*/_mergeNamespaces({
  __proto__: null
}, [account__namespace]);

var sequence = /*#__PURE__*/Object.freeze({
  __proto__: null,
  abi: abi,
  api: api,
  auth: auth,
  guard: guard,
  indexer: indexer,
  metadata: metadata,
  network: network,
  provider: provider,
  relayer: relayer,
  transactions: transactions$1,
  utils: utils,
  core: core,
  signhub: signhub,
  sessions: sessions,
  migration: migration,
  account: account,
  initWallet: provider$1.initWallet,
  getWallet: provider$1.getWallet,
  unregisterWallet: provider$1.unregisterWallet,
  SequenceProvider: provider$1.SequenceProvider,
  SequenceClient: provider$1.SequenceClient,
  SequenceSigner: provider$1.SequenceSigner
});

Object.defineProperty(exports, 'initWallet', {
  enumerable: true,
  get: function () { return provider$1.initWallet; }
});
exports.sequence = sequence;
